File: testers/db2/java/src/java/testdb/TestInsertTest.java
package testdb;

import static org.junit.Assert.*;

import java.util.Random;
import org.junit.Test;

/**
 * Tests for the TestInsert helper class that generates random test data. Tests seeded randomness,
 * customization, and foreign key handling.
 */
public class TestInsertTest {
  private final TestInsert testInsert = new TestInsert(new Random(505111434));

  @Test
  public void testCustomersInsert() {
    Db2TestHelper.run(
        c -> {
          var inserter = testInsert.Customers();

          var row = inserter.insert(c);
          assertNotNull(row);
          assertNotNull(row.customerId());
          assertNotNull(row.name());
        });
  }

  @Test
  public void testCustomersWithCustomization() {
    Db2TestHelper.run(
        c -> {
          var inserter = testInsert.Customers();

          var row = inserter.with(r -> r.withName("Custom Name")).insert(c);

          assertNotNull(row);
          assertEquals("Custom Name", row.name());
        });
  }

  @Test
  public void testOrdersWithCustomerFK() {
    Db2TestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);

          var order = testInsert.Orders(customer.customerId()).insert(c);

          assertNotNull(order);
          assertEquals(customer.customerId(), order.customerId());
        });
  }

  @Test
  public void testOrderItemsWithFK() {
    Db2TestHelper.run(
        c -> {
          var customer = testInsert.Customers().insert(c);
          var order = testInsert.Orders(customer.customerId()).insert(c);

          var orderItem = testInsert.OrderItems(order.orderId()).insert(c);

          assertNotNull(orderItem);
          assertEquals(order.orderId(), orderItem.orderId());
        });
  }

  @Test
  public void testMultipleInserts() {
    Db2TestHelper.run(
        c -> {
          var row1 = testInsert.Customers().insert(c);
          var row2 = testInsert.Customers().insert(c);
          var row3 = testInsert.Customers().insert(c);

          assertNotEquals(row1.customerId(), row2.customerId());
          assertNotEquals(row2.customerId(), row3.customerId());
          assertNotEquals(row1.customerId(), row3.customerId());
        });
  }

  @Test
  public void testInsertWithSeededRandom() {
    Db2TestHelper.run(
        c -> {
          var testInsert1 = new TestInsert(new Random(123));
          var testInsert2 = new TestInsert(new Random(123));

          var row1 = testInsert1.Customers().insert(c);
          // Use different email to avoid UNIQUE constraint violation (same seed = same email)
          var row2 = testInsert2.Customers().with(r -> r.withEmail("unique2@test.com")).insert(c);

          assertEquals(row1.name(), row2.name());
        });
  }

  @Test
  public void testChainedCustomization() {
    Db2TestHelper.run(
        c -> {
          var row =
              testInsert
                  .Customers()
                  .with(r -> r.withName("First"))
                  .with(r -> r.withEmail("first@test.com"))
                  .insert(c);

          assertEquals("First", row.name());
          assertEquals("first@test.com", row.email());
        });
  }

  @Test
  public void testIdentityAlwaysInsert() {
    Db2TestHelper.run(
        c -> {
          var row = testInsert.Db2testIdentityAlways().insert(c);

          assertNotNull(row);
          assertNotNull(row.id());
        });
  }

  @Test
  public void testIdentityDefaultInsert() {
    Db2TestHelper.run(
        c -> {
          var row = testInsert.Db2testIdentityDefault().insert(c);

          assertNotNull(row);
          assertNotNull(row.id());
        });
  }

  @Test
  public void testDistinctTypeInsert() {
    Db2TestHelper.run(
        c -> {
          var email = new EmailAddress("test@example.com");
          var row = testInsert.DistinctTypeTest(email).insert(c);

          assertNotNull(row);
          assertEquals(email, row.email());
        });
  }

  @Test
  public void testNullabilityTestInsert() {
    Db2TestHelper.run(
        c -> {
          var row = testInsert.NullabilityTest().insert(c);

          assertNotNull(row);
          assertNotNull(row.requiredCol());
        });
  }
}

File: testers/sqlserver/kotlin/src/kotlin/testdb/TestInsertTest.kt
package testdb

import org.junit.Assert.*
import org.junit.Test
import testdb.userdefined.Email
import java.math.BigDecimal
import java.util.Random

/**
 * Tests for the TestInsert helper class that generates random test data.
 */
class TestInsertTest {
    private val testInsert = TestInsert(Random(42))

    @Test
    fun testCustomersInsert() {
        SqlServerTestHelper.run { c ->
            val row = testInsert.Customers(
                name = "Test Customer",
                email = Email("test@example.com"),
                c = c
            )

            assertNotNull(row)
            assertNotNull(row.customerId)
            assertEquals("Test Customer", row.name)
            assertEquals(Email("test@example.com"), row.email)
        }
    }

    @Test
    fun testProductsInsert() {
        SqlServerTestHelper.run { c ->
            val row = testInsert.Products(
                name = "Test Product",
                price = BigDecimal("99.99"),
                c = c
            )

            assertNotNull(row)
            assertNotNull(row.productId)
            assertEquals("Test Product", row.name)
            assertEquals(0, BigDecimal("99.99").compareTo(row.price))
        }
    }

    @Test
    fun testAllScalarTypesInsert() {
        SqlServerTestHelper.run { c ->
            val row = testInsert.AllScalarTypes(c = c)

            assertNotNull(row)
            assertNotNull(row.id)
            assertNotNull(row.colRowversion)
        }
    }

    @Test
    fun testOrdersWithCustomerFK() {
        SqlServerTestHelper.run { c ->
            val customer = testInsert.Customers(
                name = "FK Customer",
                email = Email("fk@example.com"),
                c = c
            )

            val order = testInsert.Orders(
                customerId = customer.customerId,
                c = c
            )

            assertNotNull(order)
            assertEquals(customer.customerId, order.customerId)
        }
    }

    @Test
    fun testOrderItemsWithFKs() {
        SqlServerTestHelper.run { c ->
            val customer = testInsert.Customers(
                name = "OrderItem Customer",
                email = Email("orderitem@example.com"),
                c = c
            )
            val product = testInsert.Products(
                name = "OrderItem Product",
                price = BigDecimal("50.00"),
                c = c
            )
            val order = testInsert.Orders(
                customerId = customer.customerId,
                c = c
            )

            val orderItem = testInsert.OrderItems(
                orderId = order.orderId,
                productId = product.productId,
                c = c
            )

            assertNotNull(orderItem)
            assertEquals(order.orderId, orderItem.orderId)
            assertEquals(product.productId, orderItem.productId)
        }
    }

    @Test
    fun testMultipleInserts() {
        SqlServerTestHelper.run { c ->
            val row1 = testInsert.Customers(name = "Customer1", email = Email("c1@test.com"), c = c)
            val row2 = testInsert.Customers(name = "Customer2", email = Email("c2@test.com"), c = c)
            val row3 = testInsert.Customers(name = "Customer3", email = Email("c3@test.com"), c = c)

            assertNotEquals(row1.customerId, row2.customerId)
            assertNotEquals(row2.customerId, row3.customerId)
            assertNotEquals(row1.customerId, row3.customerId)
        }
    }

    @Test
    fun testInsertWithDifferentValues() {
        SqlServerTestHelper.run { c ->
            val row1 = testInsert.Customers(name = "Alpha", email = Email("alpha@test.com"), c = c)
            val row2 = testInsert.Customers(name = "Beta", email = Email("beta@test.com"), c = c)

            assertEquals("Alpha", row1.name)
            assertEquals("Beta", row2.name)
            assertEquals(Email("alpha@test.com"), row1.email)
            assertEquals(Email("beta@test.com"), row2.email)
        }
    }
}
